enterprise/server/remote_execution/action_merger/action_merger.go

```diff
@@ -204,52 +167,150 @@ func recordSubmitTimeOffsetMetric(hash map[string]string, groupIdForMetrics stri
 		Observe(float64(time.Since(submitTime).Microseconds()))
 }
 
-// Returns the execution ID of a pending execution working on the action with
-// the provided action digest, or an empty string, as well as a boolean if the
-// provided action should be run additionally in the background ("hedged"), or
-// an error if no pending execution was found.
-func FindPendingExecution(ctx context.Context, rdb redis.UniversalClient, schedulerService interfaces.SchedulerService, adResource *digest.ResourceName) (string, bool, error) {
-	if !*enableActionMerging {
-		return "", false, nil
-	}
+type Op int
+
+const (
+	// NEW signals that this is the first attempt at execution for the given
+	// action digest.
+	NEW Op = iota
+	// MERGE signals that this is not the first execution for the given action
+	// digest and the existing execution should be reused.
+	MERGE Op = iota
+	// HEDGE signals that this is not the first execution for the given action
+	// digest, but an additional execution should be run in the background.
+	HEDGE Op = iota
+)
 
+// MaybeMergeExecutions implements action merging by atomically checking if
+// there is an existing execution for the given action digest and registering
+// a new execution if not.
+//
+// The first return value is *always* a valid execution ID to be used for the
+// execution. Errors are logged, but not returned to the caller as action
+// merging is best-effort.
+//
+// The second return value indicates whether that execution ID is new (NEW), an
+// existing execution that this execution request should be merged into (MERGE),
+// or an existing execution that should be hedged with a new one (HEDGE).
+//
+// Action merging is an optimization that detects when an execution is
+// requested for an action that is in-flight, but not yet in the action cache.
+// This optimization is particularly helpful for preventing duplicate work for
+// long-running actions. It is implemented as a pair of entries in redis, one
+// from the action digest to the execution ID (the forward mapping), and one
+// from the execution ID to the action digest (the reverse mapping). These
+// entries are initially written to Redis atomically when an execution is
+// enqueued, and then rewritten (to extend the TTL) while the action is running.
+// Because the queueing mechanism isn't 100% reliable, it is possible for the
+// merging data to exist in Redis pointing to a dead execution. For this reason,
+// the TTLs are set somewhat conservatively so this problem self-heals
+// reasonably quickly.
+func MaybeMergeExecutions(ctx context.Context, rdb redis.UniversalClient, schedulerService interfaces.SchedulerService, adResource *digest.ResourceName, doNotCache bool) (string, Op) {
+	newExecutionID, err := adResource.UploadString()
+	if err != nil {
+		// adResource is always a CAS resource name, so UploadString can't
+		// return an error.
+		panic(fmt.Sprintf("UploadString unexpectedly failed on %v: %v", adResource, err))
```

Comment:
This is what motivated https://github.com/buildbuddy-io/buildbuddy/pull/8761, will update after that has been merged.